lib/pry/input_completer.rb

```diff
@@ -239,4 +227,39 @@ def build_path(input)
     end
     return path, input
   end
+
+  private
+
+  def self.fetch_obj_space_collection
```

Comment:
i still think it would be better if Pry::InputCompleter was an instance that were reused by an instance of Pry, then there wouldn't be any need for a mutex or shared state, the state being stored on the class could be stored within the instance instead. it's a bigger change though.